src/grammar/mod.rs

```diff
@@ -216,6 +217,49 @@ impl Scenario {
             })
             .collect()
     }
+
+    pub(crate) fn validate(&self) -> Result<()> {
```

Comment:
I think it is safer to do this when the contig is known. This would be in the vaftrees function then. This happens only once per contig, so the overhead is minimal.

Reply:
Then, we also do not need another version of the normalize function.

Reply:
Thanks, I didn't really know where the proper entrypoint for that is

Reply:
> Then, we also do not need another version of the normalize function.

Hm, but `normalize` returns type `NormalizedFormula`, which is problematic since we have two `NormalizedFormula`s, for which we want to check if their disjunction exists in the event definitions, but we need to normalize that disjunction, but `normalize` is only defined for `Formula`, not `NormalizedFormula`.

Perhaps instead of returning `NormalizedFormula`, return `T: Into<NormalizedFormula>` instead (and have `NormalizedFormula` implement `From<Formula>`)?
Or the other way around, i.e. make `Formula` implement `From<NormalizedFormula>` so that normalized formulas can be used as ordinary formulas again. Probably this.

Reply:
Also `vaftrees` doesn't get called if there's no rid set (for example when there are no records to be processed I guess?) not quite sure of the inner workings here. That'd also make scenario validation data dependent in some sense, I'd much rather have it validated before anything happens, i.e. upon parsing the scenario from the yaml. Though I'm not sure where to get the contigs from, then.

Reply:
> Also `vaftrees` doesn't get called if there's no rid set (for example when there are no records to be processed I guess?) not quite sure of the inner workings here. That'd also make scenario validation data dependent in some sense, I'd much rather have it validated before anything happens, i.e. upon parsing the scenario from the yaml. Though I'm not sure where to get the contigs from, then.

True, but if there is no data, it also does not really matter. I know what you mean, but I think we can live with that little caveat.